File: system-design/systems/01-rate-limiter/lab/runners/python/algorithms.py
"""Deterministic rate-limiter algorithms that emit semantic execution traces."""

from __future__ import annotations

import copy
import math
from dataclasses import dataclass
from typing import Any, Callable


class InvalidRequest(ValueError):
    """The JSON value is valid, but does not satisfy the runner contract."""


MAX_SAFE_INTEGER_MS = 9_007_199_254_740_991


def _number(value: Any, name: str, *, positive: bool = False) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise InvalidRequest(f"{name} must be a number")
    try:
        value = float(value)
    except OverflowError as failure:
        raise InvalidRequest(f"{name} must be finite") from failure
    if not math.isfinite(value) or (positive and value <= 0):
        qualifier = "a positive finite number" if positive else "finite"
        raise InvalidRequest(f"{name} must be {qualifier}")
    return value


def _quantity(value: Any, name: str) -> float:
    result = _number(value, name, positive=True)
    if result > MAX_SAFE_INTEGER_MS:
        raise InvalidRequest(
            f"{name} must be a positive finite number no greater than "
            f"{MAX_SAFE_INTEGER_MS}"
        )
    return result


def _rounded(value: float) -> int | float:
    # Quantize the actual IEEE-754 value; exact binary ties go away from zero.
    value = float(value)
    magnitude = math.floor(abs(value) * 1_000_000 + 0.5) / 1_000_000
    result = math.copysign(magnitude, value)
    if result == 0:
        return 0
    if result.is_integer():
        return int(result)
    return result


def _normalise(value: Any) -> Any:
    if isinstance(value, float):
        return _rounded(value)
    if isinstance(value, dict):
        return {key: _normalise(item) for key, item in value.items()}
    if isinstance(value, list):
        return [_normalise(item) for item in value]
    return value


@dataclass(frozen=True)
class TimelineItem:
    at_ms: float
    cost: float
    key: str


class Observer:
    def __init__(self) -> None:
        self.events: list[dict[str, Any]] = []

    def emit(
        self,
        step_id: str,
        item: TimelineItem,
        before: dict[str, Any],
        after: dict[str, Any],
        decision: dict[str, Any] | None,
        reason: str,
    ) -> None:
        self.events.append(
            {
                "seq": len(self.events) + 1,
                "stepId": step_id,
                "actor": item.key,
                "timestampMs": _rounded(item.at_ms),
                "before": _normalise(copy.deepcopy(before)),
                "after": _normalise(copy.deepcopy(after)),
                "decision": _normalise(copy.deepcopy(decision)),
                "reason": reason,
            }
        )


def _decision(
    allowed: bool,
    remaining: float,
    retry_after_ms: float,
    reset_at_ms: float,
    reason: str,
) -> dict[str, Any]:
    return {
        "allowed": allowed,
        "remaining": _rounded(max(0.0, remaining)),
        "retryAfterMs": _rounded(max(0.0, retry_after_ms)),
        "resetAtMs": _rounded(max(0.0, reset_at_ms)),
        "reason": reason,
    }


def validate_timeline(value: Any) -> list[TimelineItem]:
    if not isinstance(value, list):
        raise InvalidRequest("requestTimeline must be an array")
    if len(value) > 100:
        raise InvalidRequest("requestTimeline must contain at most 100 items")
    result: list[TimelineItem] = []
    previous = -1.0
    for index, raw in enumerate(value):
        if not isinstance(raw, dict):
            raise InvalidRequest(f"requestTimeline[{index}] must be an object")
        at_ms = _number(raw.get("atMs"), f"requestTimeline[{index}].atMs")
        cost = _quantity(raw.get("cost"), f"requestTimeline[{index}].cost")
        key = raw.get("key")
        if (
            not at_ms.is_integer()
            or at_ms < 0
            or at_ms > MAX_SAFE_INTEGER_MS
            or at_ms < previous
        ):
            raise InvalidRequest(
                "requestTimeline atMs must be non-negative, non-decreasing safe integer milliseconds"
            )
        try:
            key_bytes = key.encode("utf-8") if isinstance(key, str) else b""
        except UnicodeEncodeError:
            key_bytes = b""
        if not isinstance(key, str) or not key or not key_bytes or len(key_bytes) > 128:
            raise InvalidRequest(
                f"requestTimeline[{index}].key must be a non-empty UTF-8 string "
                "of at most 128 bytes"
            )
        result.append(TimelineItem(at_ms, cost, key))
        previous = at_ms
    return result


def _window_config(config: Any) -> tuple[float, float]:
    if not isinstance(config, dict):
        raise InvalidRequest("config must be an object")
    limit = _quantity(config.get("limit"), "config.limit")
    window_ms = _number(config.get("windowMs"), "config.windowMs", positive=True)
    if not window_ms.is_integer() or window_ms > MAX_SAFE_INTEGER_MS:
        raise InvalidRequest(
            "config.windowMs must be a positive safe integer in milliseconds"
        )
    return limit, window_ms


def _bucket_config(config: Any) -> tuple[float, float]:
    if not isinstance(config, dict):
        raise InvalidRequest("config must be an object")
    capacity = _quantity(config.get("capacity"), "config.capacity")
    rate = _quantity(config.get("ratePerSecond"), "config.ratePerSecond")
    rate_per_ms = rate / 1000.0
    if rate_per_ms <= 0 or capacity > MAX_SAFE_INTEGER_MS * rate_per_ms:
        raise InvalidRequest(
            "config capacity and ratePerSecond must yield a finite maximum recovery "
            f"time no greater than {MAX_SAFE_INTEGER_MS} milliseconds"
        )
    maximum_recovery_ms = capacity / rate_per_ms
    if not math.isfinite(maximum_recovery_ms) or maximum_recovery_ms > MAX_SAFE_INTEGER_MS:
        raise InvalidRequest(
            "config capacity and ratePerSecond must yield a finite maximum recovery "
            f"time no greater than {MAX_SAFE_INTEGER_MS} milliseconds"
        )
    return capacity, rate


def run_fixed_window(config: Any, timeline: list[TimelineItem], observer: Observer) -> list[dict[str, Any]]:
    limit, window_ms = _window_config(config)
    states: dict[str, dict[str, float]] = {}
    decisions = []
    for item in timeline:
        window_start = math.floor(item.at_ms / window_ms) * window_ms
        state = states.setdefault(item.key, {"windowStartMs": window_start, "count": 0.0})
        before = copy.deepcopy(state)
        # @step:fixed.locate-window
        if state["windowStartMs"] != window_start:
            state["windowStartMs"] = window_start
            state["count"] = 0.0
        observer.emit("fixed.locate-window", item, before, state, None, "window_selected")

        decision_before = copy.deepcopy(state)
        allowed = state["count"] + item.cost <= limit
        if allowed:
            state["count"] += item.cost
        reason = "within_limit" if allowed else (
            "cost_exceeds_limit" if item.cost > limit else "limit_exceeded"
        )
        reset_at = window_start + window_ms
        # @step:fixed.decision
        decision = _decision(
            allowed,
            limit - state["count"],
            0 if allowed or item.cost > limit else reset_at - item.at_ms,
            reset_at,
            reason,
        )
        observer.emit("fixed.decision", item, decision_before, state, decision, reason)
        decisions.append(decision)
    return decisions


def run_sliding_log(config: Any, timeline: list[TimelineItem], observer: Observer) -> list[dict[str, Any]]:
    limit, window_ms = _window_config(config)
    states: dict[str, list[dict[str, float]]] = {}
    decisions = []
    for item in timeline:
        entries = states.setdefault(item.key, [])
        before = {"entries": copy.deepcopy(entries), "used": sum(entry["cost"] for entry in entries)}
        cutoff = item.at_ms - window_ms
        # @step:sliding-log.evict
        entries[:] = [entry for entry in entries if entry["atMs"] > cutoff]
        used = sum(entry["cost"] for entry in entries)
        after_evict = {"entries": copy.deepcopy(entries), "used": used}
        observer.emit("sliding-log.evict", item, before, after_evict, None, "expired_entries_removed")

        allowed = used + item.cost <= limit
        if allowed:
            entries.append({"atMs": item.at_ms, "cost": item.cost})
            used += item.cost
        reason = "within_limit" if allowed else (
            "cost_exceeds_limit" if item.cost > limit else "limit_exceeded"
        )
        retry_after = 0.0
        if not allowed and item.cost <= limit:
            required_release = used + item.cost - limit
            released = 0.0
            for entry in entries:
                released += entry["cost"]
                if released + 1e-9 >= required_release:
                    retry_after = max(0.0, entry["atMs"] + window_ms - item.at_ms)
                    break
        reset_at = item.at_ms + retry_after if not allowed else (
            entries[0]["atMs"] + window_ms if entries else item.at_ms
        )
        after = {"entries": copy.deepcopy(entries), "used": used}
        # @step:sliding-log.decision
        decision = _decision(allowed, limit - used, retry_after, reset_at, reason)
        observer.emit("sliding-log.decision", item, after_evict, after, decision, reason)
        decisions.append(decision)
    return decisions


def run_sliding_counter(config: Any, timeline: list[TimelineItem], observer: Observer) -> list[dict[str, Any]]:
    limit, window_ms = _window_config(config)
    states: dict[str, dict[str, float]] = {}
    decisions = []
    for item in timeline:
        current_start = math.floor(item.at_ms / window_ms) * window_ms
        state = states.setdefault(
            item.key,
            {"currentWindowStartMs": current_start, "currentCount": 0.0, "previousCount": 0.0},
        )
        before = copy.deepcopy(state)
        windows_elapsed = int((current_start - state["currentWindowStartMs"]) / window_ms)
        # @step:sliding-counter.rotate
        if windows_elapsed == 1:
            state["previousCount"] = state["currentCount"]
            state["currentCount"] = 0.0
            state["currentWindowStartMs"] = current_start
        elif windows_elapsed > 1:
            state["previousCount"] = 0.0
            state["currentCount"] = 0.0
            state["currentWindowStartMs"] = current_start
        observer.emit("sliding-counter.rotate", item, before, state, None, "windows_rotated")

        elapsed = item.at_ms - current_start
        previous_weight = max(0.0, 1.0 - elapsed / window_ms)
        # @step:sliding-counter.estimate
        estimated = state["currentCount"] + state["previousCount"] * previous_weight
        estimate_state = {**state, "previousWeight": previous_weight, "estimatedCount": estimated}
        observer.emit("sliding-counter.estimate", item, state, estimate_state, None, "weighted_count_estimated")

        allowed = estimated + item.cost <= limit
        if allowed:
            state["currentCount"] += item.cost
            estimated += item.cost
        reason = "within_limit" if allowed else (
            "cost_exceeds_limit" if item.cost > limit else "limit_exceeded"
        )
        reset_at = current_start + window_ms
        retry_after = 0.0
        if not allowed and item.cost <= limit:
            excess = max(0.0, estimated + item.cost - limit)
            until_boundary = max(0.0, current_start + window_ms - item.at_ms)
            previous_count = state["previousCount"]
            current_count = state["currentCount"]
            if current_count + item.cost <= limit and previous_count > 0:
                retry_after = min(until_boundary, excess * window_ms / previous_count)
            else:
                retry_after = until_boundary
                if current_count > 0:
                    retry_after += max(
                        0.0,
                        (current_count + item.cost - limit) * window_ms / current_count,
                    )
            reset_at = item.at_ms + retry_after
        after = {**state, "previousWeight": previous_weight, "estimatedCount": estimated}
        # @step:sliding-counter.decision
        decision = _decision(
            allowed,
            limit - estimated,
            retry_after,
            reset_at,
            reason,
        )
        observer.emit("sliding-counter.decision", item, estimate_state, after, decision, reason)
        decisions.append(decision)
    return decisions


def run_token_bucket(config: Any, timeline: list[TimelineItem], observer: Observer) -> list[dict[str, Any]]:
    capacity, rate = _bucket_config(config)
    states: dict[str, dict[str, float]] = {}
    decisions = []
    for item in timeline:
        state = states.setdefault(item.key, {"tokens": capacity, "lastRefillMs": item.at_ms})
        before = copy.deepcopy(state)
        elapsed = max(0.0, item.at_ms - state["lastRefillMs"])
        # @step:token.refill
        state["tokens"] = min(capacity, state["tokens"] + elapsed * rate / 1000.0)
        state["lastRefillMs"] = item.at_ms
        observer.emit("token.refill", item, before, state, None, "tokens_refilled")

        decision_before = copy.deepcopy(state)
        allowed = state["tokens"] + 1e-9 >= item.cost
        if allowed:
            state["tokens"] -= item.cost
        reason = "token_available" if allowed else (
            "cost_exceeds_capacity" if item.cost > capacity else "insufficient_tokens"
        )
        retry_after = (
            0.0
            if allowed or item.cost > capacity
            else (item.cost - state["tokens"]) / rate * 1000.0
        )
        reset_at = item.at_ms + (capacity - state["tokens"]) / rate * 1000.0
        # @step:token.decision
        decision = _decision(allowed, state["tokens"], retry_after, reset_at, reason)
        observer.emit("token.decision", item, decision_before, state, decision, reason)
        decisions.append(decision)
    return decisions


def run_leaky_bucket(config: Any, timeline: list[TimelineItem], observer: Observer) -> list[dict[str, Any]]:
    capacity, rate = _bucket_config(config)
    states: dict[str, dict[str, float]] = {}
    decisions = []
    for item in timeline:
        state = states.setdefault(item.key, {"water": 0.0, "lastLeakMs": item.at_ms})
        before = copy.deepcopy(state)
        elapsed = max(0.0, item.at_ms - state["lastLeakMs"])
        # @step:leaky.drain
        state["water"] = max(0.0, state["water"] - elapsed * rate / 1000.0)
        state["lastLeakMs"] = item.at_ms
        observer.emit("leaky.drain", item, before, state, None, "queued_work_drained")

        decision_before = copy.deepcopy(state)
        allowed = state["water"] + item.cost <= capacity + 1e-9
        if allowed:
            state["water"] += item.cost
        reason = "queue_has_capacity" if allowed else (
            "cost_exceeds_capacity" if item.cost > capacity else "queue_full"
        )
        retry_after = (
            0.0
            if allowed or item.cost > capacity
            else (state["water"] + item.cost - capacity) / rate * 1000.0
        )
        reset_at = item.at_ms + state["water"] / rate * 1000.0
        # @step:leaky.decision
        decision = _decision(allowed, capacity - state["water"], retry_after, reset_at, reason)
        observer.emit("leaky.decision", item, decision_before, state, decision, reason)
        decisions.append(decision)
    return decisions


RUNNERS: dict[str, Callable[[Any, list[TimelineItem], Observer], list[dict[str, Any]]]] = {
    "fixed-window": run_fixed_window,
    "sliding-window-log": run_sliding_log,
    "sliding-window-counter": run_sliding_counter,
    "token-bucket": run_token_bucket,
    "leaky-bucket": run_leaky_bucket,
}


def run_request(request: Any) -> dict[str, Any]:
    if not isinstance(request, dict):
        raise InvalidRequest("request must be an object")
    algorithm = request.get("algorithm")
    if algorithm not in RUNNERS:
        raise InvalidRequest(f"unsupported algorithm: {algorithm!r}")
    timeline = validate_timeline(request.get("requestTimeline"))
    observer = Observer()
    decisions = RUNNERS[algorithm](request.get("config"), timeline, observer)
    return {"events": observer.events, "decisions": decisions}

File: system-design/systems/01-rate-limiter/lab/runners/js/algorithms.mjs
/**
 * Deterministic rate-limiter algorithms with observer-driven semantic traces.
 *
 * A synchronous allow path finishes in one event-loop turn and does not interleave with another
 * synchronous call. Read-modify-write split across await points can still race. Cross-process and
 * distributed quotas therefore require Redis atomic commands/Lua; the event loop is not a lock.
 */

export class InvalidRequest extends Error {}

function number(value, name, { positive = false } = {}) {
  if (typeof value !== "number" || !Number.isFinite(value) || (positive && value <= 0)) {
    throw new InvalidRequest(`${name} must be ${positive ? "a positive finite number" : "finite"}`);
  }
  return value;
}

function quantity(value, name) {
  const result = number(value, name, { positive: true });
  if (result > Number.MAX_SAFE_INTEGER) {
    throw new InvalidRequest(
      `${name} must be a positive finite number no greater than ${Number.MAX_SAFE_INTEGER}`,
    );
  }
  return result;
}

function rounded(value) {
  // Quantize the actual IEEE-754 value; exact binary ties go away from zero.
  const magnitude = Math.floor(Math.abs(value) * 1_000_000 + 0.5) / 1_000_000;
  const result = Math.sign(value) * magnitude;
  if (Object.is(result, -0) || result === 0) return 0;
  return Number.isInteger(result) ? Math.trunc(result) : result;
}

function normalise(value) {
  if (typeof value === "number") return rounded(value);
  if (Array.isArray(value)) return value.map(normalise);
  if (value && typeof value === "object") {
    return Object.fromEntries(Object.entries(value).map(([key, item]) => [key, normalise(item)]));
  }
  return value;
}

function snapshot(value) {
  return normalise(structuredClone(value));
}

class Observer {
  constructor() {
    this.events = [];
  }

  emit(stepId, item, before, after, decision, reason) {
    this.events.push({
      seq: this.events.length + 1,
      stepId,
      actor: item.key,
      timestampMs: rounded(item.atMs),
      before: snapshot(before),
      after: snapshot(after),
      decision: decision === null ? null : snapshot(decision),
      reason,
    });
  }
}

function decision(allowed, remaining, retryAfterMs, resetAtMs, reason) {
  return {
    allowed,
    remaining: rounded(Math.max(0, remaining)),
    retryAfterMs: rounded(Math.max(0, retryAfterMs)),
    resetAtMs: rounded(Math.max(0, resetAtMs)),
    reason,
  };
}

function validateTimeline(value) {
  if (!Array.isArray(value)) throw new InvalidRequest("requestTimeline must be an array");
  if (value.length > 100) throw new InvalidRequest("requestTimeline must contain at most 100 items");
  let previous = -1;
  return value.map((raw, index) => {
    if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
      throw new InvalidRequest(`requestTimeline[${index}] must be an object`);
    }
    const atMs = number(raw.atMs, `requestTimeline[${index}].atMs`);
    const cost = quantity(raw.cost, `requestTimeline[${index}].cost`);
    const key = raw.key;
    if (!Number.isSafeInteger(atMs) || atMs < 0 || atMs < previous) {
      throw new InvalidRequest(
        "requestTimeline atMs must be non-negative, non-decreasing safe integer milliseconds",
      );
    }
    if (
      typeof key !== "string"
      || key.length === 0
      || new TextEncoder().encode(key).byteLength > 128
    ) {
      throw new InvalidRequest(
        `requestTimeline[${index}].key must be a non-empty UTF-8 string of at most 128 bytes`,
      );
    }
    previous = atMs;
    return { atMs, cost, key };
  });
}

function windowConfig(config) {
  if (!config || typeof config !== "object" || Array.isArray(config)) {
    throw new InvalidRequest("config must be an object");
  }
  const limit = quantity(config.limit, "config.limit");
  const windowMs = number(config.windowMs, "config.windowMs", { positive: true });
  if (!Number.isSafeInteger(windowMs)) {
    throw new InvalidRequest("config.windowMs must be a positive safe integer in milliseconds");
  }
  return { limit, windowMs };
}

function bucketConfig(config) {
  if (!config || typeof config !== "object" || Array.isArray(config)) {
    throw new InvalidRequest("config must be an object");
  }
  const capacity = quantity(config.capacity, "config.capacity");
  const rate = quantity(config.ratePerSecond, "config.ratePerSecond");
  const maximumRecoveryMs = capacity / (rate / 1000);
  if (!Number.isFinite(maximumRecoveryMs) || maximumRecoveryMs > Number.MAX_SAFE_INTEGER) {
    throw new InvalidRequest(
      "config capacity and ratePerSecond must yield a finite maximum recovery "
      + `time no greater than ${Number.MAX_SAFE_INTEGER} milliseconds`,
    );
  }
  return { capacity, rate };
}

function runFixedWindow(config, timeline, observer) {
  const { limit, windowMs } = windowConfig(config);
  const states = new Map();
  const decisions = [];
  for (const item of timeline) {
    const windowStartMs = Math.floor(item.atMs / windowMs) * windowMs;
    const state = states.get(item.key) ?? { windowStartMs, count: 0 };
    states.set(item.key, state);
    const before = structuredClone(state);
    // @step:fixed.locate-window
    if (state.windowStartMs !== windowStartMs) {
      state.windowStartMs = windowStartMs;
      state.count = 0;
    }
    observer.emit("fixed.locate-window", item, before, state, null, "window_selected");

    const decisionBefore = structuredClone(state);
    const allowed = state.count + item.cost <= limit;
    if (allowed) state.count += item.cost;
    const reason = allowed ? "within_limit" : (item.cost > limit ? "cost_exceeds_limit" : "limit_exceeded");
    const resetAtMs = windowStartMs + windowMs;
    // @step:fixed.decision
    const result = decision(
      allowed,
      limit - state.count,
      allowed || item.cost > limit ? 0 : resetAtMs - item.atMs,
      resetAtMs,
      reason,
    );
    observer.emit("fixed.decision", item, decisionBefore, state, result, reason);
    decisions.push(result);
  }
  return decisions;
}

function runSlidingLog(config, timeline, observer) {
  const { limit, windowMs } = windowConfig(config);
  const states = new Map();
  const decisions = [];
  for (const item of timeline) {
    const entries = states.get(item.key) ?? [];
    states.set(item.key, entries);
    const before = { entries: structuredClone(entries), used: entries.reduce((sum, entry) => sum + entry.cost, 0) };
    const cutoff = item.atMs - windowMs;
    // @step:sliding-log.evict
    while (entries.length > 0 && entries[0].atMs <= cutoff) entries.shift();
    let used = entries.reduce((sum, entry) => sum + entry.cost, 0);
    const afterEvict = { entries: structuredClone(entries), used };
    observer.emit("sliding-log.evict", item, before, afterEvict, null, "expired_entries_removed");

    const allowed = used + item.cost <= limit;
    if (allowed) {
      entries.push({ atMs: item.atMs, cost: item.cost });
      used += item.cost;
    }
    const reason = allowed ? "within_limit" : (item.cost > limit ? "cost_exceeds_limit" : "limit_exceeded");
    let retryAfterMs = 0;
    if (!allowed && item.cost <= limit) {
      const requiredRelease = used + item.cost - limit;
      let released = 0;
      for (const entry of entries) {
        released += entry.cost;
        if (released + 1e-9 >= requiredRelease) {
          retryAfterMs = Math.max(0, entry.atMs + windowMs - item.atMs);
          break;
        }
      }
    }
    const resetAtMs = allowed
      ? (entries[0]?.atMs ?? item.atMs - windowMs) + windowMs
      : item.atMs + retryAfterMs;
    const after = { entries: structuredClone(entries), used };
    // @step:sliding-log.decision
    const result = decision(allowed, limit - used, retryAfterMs, resetAtMs, reason);
    observer.emit("sliding-log.decision", item, afterEvict, after, result, reason);
    decisions.push(result);
  }
  return decisions;
}

function runSlidingCounter(config, timeline, observer) {
  const { limit, windowMs } = windowConfig(config);
  const states = new Map();
  const decisions = [];
  for (const item of timeline) {
    const currentWindowStartMs = Math.floor(item.atMs / windowMs) * windowMs;
    const state = states.get(item.key) ?? {
      currentWindowStartMs,
      currentCount: 0,
      previousCount: 0,
    };
    states.set(item.key, state);
    const before = structuredClone(state);
    const windowsElapsed = Math.floor((currentWindowStartMs - state.currentWindowStartMs) / windowMs);
    // @step:sliding-counter.rotate
    if (windowsElapsed === 1) {
      state.previousCount = state.currentCount;
      state.currentCount = 0;
      state.currentWindowStartMs = currentWindowStartMs;
    } else if (windowsElapsed > 1) {
      state.previousCount = 0;
      state.currentCount = 0;
      state.currentWindowStartMs = currentWindowStartMs;
    }
    observer.emit("sliding-counter.rotate", item, before, state, null, "windows_rotated");

    const elapsed = item.atMs - currentWindowStartMs;
    const previousWeight = Math.max(0, 1 - elapsed / windowMs);
    // @step:sliding-counter.estimate
    let estimatedCount = state.currentCount + state.previousCount * previousWeight;
    const estimateState = { ...state, previousWeight, estimatedCount };
    observer.emit("sliding-counter.estimate", item, state, estimateState, null, "weighted_count_estimated");

    const allowed = estimatedCount + item.cost <= limit;
    if (allowed) {
      state.currentCount += item.cost;
      estimatedCount += item.cost;
    }
    const reason = allowed ? "within_limit" : (item.cost > limit ? "cost_exceeds_limit" : "limit_exceeded");
    let resetAtMs = currentWindowStartMs + windowMs;
    let retryAfterMs = 0;
    if (!allowed && item.cost <= limit) {
      const excess = Math.max(0, estimatedCount + item.cost - limit);
      const untilBoundary = Math.max(0, currentWindowStartMs + windowMs - item.atMs);
      if (state.currentCount + item.cost <= limit && state.previousCount > 0) {
        retryAfterMs = Math.min(untilBoundary, (excess * windowMs) / state.previousCount);
      } else {
        retryAfterMs = untilBoundary;
        if (state.currentCount > 0) {
          retryAfterMs += Math.max(
            0,
            ((state.currentCount + item.cost - limit) * windowMs) / state.currentCount,
          );
        }
      }
      resetAtMs = item.atMs + retryAfterMs;
    }
    const after = { ...state, previousWeight, estimatedCount };
    // @step:sliding-counter.decision
    const result = decision(
      allowed,
      limit - estimatedCount,
      retryAfterMs,
      resetAtMs,
      reason,
    );
    observer.emit("sliding-counter.decision", item, estimateState, after, result, reason);
    decisions.push(result);
  }
  return decisions;
}

function runTokenBucket(config, timeline, observer) {
  const { capacity, rate } = bucketConfig(config);
  const states = new Map();
  const decisions = [];
  for (const item of timeline) {
    const state = states.get(item.key) ?? { tokens: capacity, lastRefillMs: item.atMs };
    states.set(item.key, state);
    const before = structuredClone(state);
    const elapsed = Math.max(0, item.atMs - state.lastRefillMs);
    // @step:token.refill
    state.tokens = Math.min(capacity, state.tokens + (elapsed * rate) / 1000);
    state.lastRefillMs = item.atMs;
    observer.emit("token.refill", item, before, state, null, "tokens_refilled");

    const decisionBefore = structuredClone(state);
    const allowed = state.tokens + 1e-9 >= item.cost;
    if (allowed) state.tokens -= item.cost;
    const reason = allowed
      ? "token_available"
      : (item.cost > capacity ? "cost_exceeds_capacity" : "insufficient_tokens");
    const retryAfterMs = allowed || item.cost > capacity
      ? 0
      : ((item.cost - state.tokens) / rate) * 1000;
    const resetAtMs = item.atMs + ((capacity - state.tokens) / rate) * 1000;
    // @step:token.decision
    const result = decision(allowed, state.tokens, retryAfterMs, resetAtMs, reason);
    observer.emit("token.decision", item, decisionBefore, state, result, reason);
    decisions.push(result);
  }
  return decisions;
}

function runLeakyBucket(config, timeline, observer) {
  const { capacity, rate } = bucketConfig(config);
  const states = new Map();
  const decisions = [];
  for (const item of timeline) {
    const state = states.get(item.key) ?? { water: 0, lastLeakMs: item.atMs };
    states.set(item.key, state);
    const before = structuredClone(state);
    const elapsed = Math.max(0, item.atMs - state.lastLeakMs);
    // @step:leaky.drain
    state.water = Math.max(0, state.water - (elapsed * rate) / 1000);
    state.lastLeakMs = item.atMs;
    observer.emit("leaky.drain", item, before, state, null, "queued_work_drained");

    const decisionBefore = structuredClone(state);
    const allowed = state.water + item.cost <= capacity + 1e-9;
    if (allowed) state.water += item.cost;
    const reason = allowed
      ? "queue_has_capacity"
      : (item.cost > capacity ? "cost_exceeds_capacity" : "queue_full");
    const retryAfterMs = allowed || item.cost > capacity
      ? 0
      : ((state.water + item.cost - capacity) / rate) * 1000;
    const resetAtMs = item.atMs + (state.water / rate) * 1000;
    // @step:leaky.decision
    const result = decision(allowed, capacity - state.water, retryAfterMs, resetAtMs, reason);
    observer.emit("leaky.decision", item, decisionBefore, state, result, reason);
    decisions.push(result);
  }
  return decisions;
}

const runners = new Map([
  ["fixed-window", runFixedWindow],
  ["sliding-window-log", runSlidingLog],
  ["sliding-window-counter", runSlidingCounter],
  ["token-bucket", runTokenBucket],
  ["leaky-bucket", runLeakyBucket],
]);

export function runRequest(request) {
  if (!request || typeof request !== "object" || Array.isArray(request)) {
    throw new InvalidRequest("request must be an object");
  }
  const runner = runners.get(request.algorithm);
  if (!runner) throw new InvalidRequest(`unsupported algorithm: ${String(request.algorithm)}`);
  const timeline = validateTimeline(request.requestTimeline);
  const observer = new Observer();
  const decisions = runner(request.config, timeline, observer);
  return { events: observer.events, decisions };
}
